tbot/machine/board/zephyr.py

```diff
@@ -0,0 +1,208 @@
+
+
+class ZephyrShell(shell.Shell, ZephyrStartup):
+    """
+    Zephyr shell.
+
+    The interface of this shell is really simple. It provides
+
+    - :py:meth:`ze.exec() <tbot.machine.board.ZephyrShell.exec>` - Run command
+      and return output.
+    - :py:meth:`ze.interactive() <tbot.machine.board.ZephyrShell.interactive>` -
+      Start an interactive session for this machine.
+    """
+
+    prompt: typing.Union[str, bytes] = "shell> "
+    """
+    Prompt which was configured for Zephyr.
+
+    Commonly ``"shell> "``.
+
+    .. warning::
+
+        **Don't forget the trailing space, if your prompt has one!**
+    """
+
+    bootlog: str
+    """Transcript of console output during boot."""
+
+    @contextlib.contextmanager
+    def _init_shell(self) -> typing.Iterator:
+        with self._zephyr_startup_event() as ev, self.ch.with_stream(ev):
+            self.ch.prompt = (
+                self.prompt.encode("utf-8")
+                if isinstance(self.prompt, str)
+                else self.prompt
+            )
+
+            # Set a blacklist of control characters.  These characters were
+            # copied over from the U-Boot board.
+            self.ch._write_blacklist = [
+                0x00,  # NUL  | Null
+                0x01,  # SOH  | Start of Heading
+                0x02,  # STX  | Start of Text
+                0x03,  # ETX  | End of Text / Interrupt
+                0x04,  # EOT  | End of Transmission
+                0x05,  # ENQ  | Enquiry
+                0x06,  # ACK  | Acknowledge
+                0x07,  # BEL  | Bell, Alert
+                0x08,  # BS   | Backspace
+                0x09,  # HT   | Character Tabulation, Horizontal Tabulation
+                0x0B,  # VT   | Line Tabulation, Vertical Tabulation
+                0x0C,  # FF   | Form Feed
+                0x0E,  # SO   | Shift Out
+                0x0F,  # SI   | Shift In
+                0x10,  # DLE  | Data Link Escape
+                0x11,  # DC1  | Device Control One (XON)
+                0x12,  # DC2  | Device Control Two
+                0x13,  # DC3  | Device Control Three (XOFF)
+                0x14,  # DC4  | Device Control Four
+                0x15,  # NAK  | Negative Acknowledge
+                0x16,  # SYN  | Synchronous Idle
+                0x17,  # ETB  | End of Transmission Block
+                0x18,  # CAN  | Cancel
+                0x1A,  # SUB  | Substitute / Suspend Process
+                0x1B,  # ESC  | Escape
+                0x1C,  # FS   | File Separator
+                0x7F,  # DEL  | Delete
+            ]
+
+            while True:
+                if self.boot_timeout is not None:
+                    assert self._timeout_start is not None
+                    if (time.monotonic() - self._timeout_start) > self.boot_timeout:
+                        raise TimeoutError("Zephyr did not reach shell in time")
+                try:
+                    self.ch.read_until_prompt(timeout=0.2)
```

Comment:
I have an issue with that code: my testcases succeed 95% of the time, but miss the prompt every so often which results in a TimeoutError exception. Adding `print(buf)` in `read_until_prompt()` just before [1] shows that `buf` is truncated when this happens. Here's an sample output of such a timeout, looking for `b'# MY_PROMPT'`:
```
[...]
bytearray(b'[... hundreds\r\n of\r\n bytes]\r\n # MY_P')                                                                                 
Ebytearray(b'R')                                                                                                                                                                                                                        
SUbytearray(b'ROM')                                                                                                                                                                                                                     
Lbytearray(b'ROMP')                                                                                                                                                                                                                     
bytearray(b'ROMPT\r')                                                                                                                                                                                                                   
│   │   │    <>                                                                                                     
bytearray(b'ROMPT\r\n')                                                                                                                                                                                                                 
│   │   ├─Result: TIMEOUT                                                                                           
│   │   └─Fail. (46.198s)                                                                                                                                                                                                               
```
The problem disappears when I set this timeout to something big (10 seconds) (in my case, boot takes around 2 seconds to get to the prompt). As this timeout value is given to `read_iter()` in `channel.py`, I believe that based on luck, a timeout can occur right during the display of the prompt. This splits the buffer in the middle of the prompt and we therefore miss it.

This problem seems to be present for U-Boot boards too (I took that code from there), but perhaps boot time is far less than 200ms in most cases, so it doesn't happen in practice. I'm not sure about how Linux boards handle that part.

I don't know what would be the best way to handle this. Perhaps read_iter() could store the buffer and restore it after a Timeout exception? I'm new to that Iterator/yield() stuff, so bear with me here.

[1] https://github.com/Rahix/tbot/blob/master/tbot/machine/channel/channel.py#L881

Reply:
I see that ^C is sent on timeout, which displays the prompt again. So U-Boot has probably no issue with that, I guess you can dismiss that comment :).

Reply:
Yeah, the idea is that e.g. if the board is already turned on, the prompt won't re-appear on its own when attaching to the serial console.  So I send `^C` which makes U-Boot show the prompt again and tbot can then continue.

I guess the same would be useful for Zephyr but as you noticed `^C` does not work, we'll have to find another mechanism ...  Using `\n` is a bit dangerous: For example, there might be a half-written command currently waiting on the console before tbot attaches.  Sending `\n` would then execute it which should be avoided.  Is there any way we could prevent this?  E.g. is there any key sequence we can use to clear everything currently written after the prompt (short of sending a bunch of backspaces :p)?

Reply:
Sorry, I didn't refresh my page and missed your reply.

>  Using \n is a bit dangerous

Oh yeah, I didn't think about that. I was mainly concerned by the "prompt split in two" issue.

>  E.g. is there any key sequence we can use to clear everything currently written after the prompt (short of sending a bunch of backspaces :p)?

So it looks like latest Zephyr versions handle ctrl keys [1] but I've never tried. Looks like it should work though!

[1] https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/shell/shell.c#L822